### ISSUE TITLE
[Question] details about ZMQ + FCI for real Franka

### ISSUE THREAD

**Issue by chauvinSimon** [NVIDIA-Omniverse/Orbit#49]
> "we use **`ZMQ` to send joint commands from `ORBIT` to a computer running the real-time kernel for `Franka Emika` robot**. To abide by the real-time safety constraints, we use a quintic interpolator to upsample the `60 Hz` joint commands from the simulator to `1000 Hz` for execution on the robot" ([`orbit` paper](https://arxiv.org/pdf/2301.04195.pdf))

---

This sound **very efficient** to overcome the **impossibility to use nvidia-drivers on the PC that talks to the robot** ([due to the `PREEMPT_RT` kernel](https://frankaemika.github.io/docs/installation_linux.html#setting-up-the-real-time-kernel))!

I would have two series of questions (unfortunately I could not find any additional details in the documentation / paper):
1. regarding the **`zmq` interface**:
   1. **what serialization** is used (maybe [`msgpack`](https://msgpack.org/) or [`protobuf`](https://github.com/protocolbuffers/protobuf))?
   2. :octocat: is the **code available** somewhere?

2. regarding the computer running the RT kernel for the Franka Emika robot:
   1. it is probably using `FCI` via [`libfranka`](https://frankaemika.github.io/docs/libfranka.html), isn't it?
   2. is the **target `joint position`** (called `qc` on [this `FCI` figure](https://frankaemika.github.io/docs/libfranka.html#under-the-hood)) passed to a **`joint position motion generator`**, which is then processed by one of the **`FCI` internal controllers, e.g. the `joint impedance controller`**, as explained [here](https://frankaemika.github.io/docs/libfranka.html)? _(green path on the attached image)_
   3. or are the **target `torques`** (called `to-d` on [this `FCI` figure](https://frankaemika.github.io/docs/libfranka.html#under-the-hood)) directly processed by **an `external controller`**? _(purple path on the attached image)_
   4. on the [`fig.9`](https://www.semanticscholar.org/paper/ORBIT%3A-A-Unified-Simulation-Framework-for-Robot-Mittal-Yu/60219c9da45ec42560ade9f4adf792776ecc763a/figure/7) the target `joint position` issued at `1kHz` is converted to `torques` by a `PD` controller.
      1. is this `PD` controller the **built-in one of `FCI`**?
      1. is this `PD` controller also used in simulation during training _(I do not think so)_? And if yes are the **`PD` parameters identical** in **simulation and on the real robot**?
   5. :octocat: is the **code available** somewhere?

---
---

![20230404_franka_zmq](https://user-images.githubusercontent.com/34301767/229797614-87d4a81c-7edb-4da9-b233-7e58a2192437.png)

---
---

![fci](https://user-images.githubusercontent.com/34301767/229803435-4e65fb65-5bd5-487c-a5e2-21f9039c3f8f.png)

**Comment by ritviksingh9** [comment]
1. We do not use any serialisation library. At max we only sent 23 (7+16) doubles which is already a fairly compressed representation to send over machines connected on the same local network. As such there was no need serialise values.
2. Yes, we use `libfranka`. We do not pass the joint position targets to one of the joint position motion generators provided by `libfranka`. Instead, we created our own `TorqueGenerator` which contains our own PD controller. This outputs joint torques that are sent to the Franka. So to answer your question, we use the purple path to create an external controller, and the PD controller is not the built-in one in the FCI, it's our own PD controller for which we spent some time tuning the gains. The PD controller is not identical to what is used in training in sim.

The real-robot code is currently not available publicly, but it is something that we are currently working on.
